File: Fibonacci Heaps/fibonacci_heap.py
# File: fibonacci_heap.py
# Program: Create a Fibonacci Heap and insert nodes

class Node:
    def __init__(self, key):
        self.key = key
        self.degree = 0
        self.parent = None
        self.child = None
        self.left = self
        self.right = self


class FibonacciHeap:

    def __init__(self):
        self.min = None

    def insert(self, key):

        node = Node(key)

        if self.min is None:
            self.min = node
        else:

            node.left = self.min
            node.right = self.min.right
            self.min.right.left = node
            self.min.right = node

            if node.key < self.min.key:
                self.min = node

        print("Inserted node", key)

    def display(self):

        if self.min is None:
            return

        temp = self.min
        print("Root List:", end=" ")

        while True:
            print(temp.key, end=" ")
            temp = temp.right
            if temp == self.min:
                break

        print()


heap = FibonacciHeap()

heap.insert(10)
heap.insert(3)
heap.insert(15)
heap.insert(6)

heap.display()

File: Fibonacci Heaps/fibonacci_heap.c
/* File: fibonacci_heap.c
   Program: Create a Fibonacci Heap and insert nodes
*/

#include <stdio.h>
#include <stdlib.h>

typedef struct node {
    int key;
    int degree;
    struct node *parent;
    struct node *child;
    struct node *left;
    struct node *right;
} Node;

Node *min = NULL;

Node* createNode(int key)
{
    Node *n = (Node*)malloc(sizeof(Node));
    n->key = key;
    n->degree = 0;
    n->parent = NULL;
    n->child = NULL;
    n->left = n;
    n->right = n;
    return n;
}

void insert(int key)
{
    Node *n = createNode(key);

    if (min == NULL)
        min = n;
    else
    {
        n->left = min;
        n->right = min->right;
        min->right->left = n;
        min->right = n;

        if (n->key < min->key)
            min = n;
    }

    printf("Inserted node %d\n", key);
}

void display()
{
    Node *temp = min;
    if (temp == NULL)
        return;

    printf("Root List: ");
    do
    {
        printf("%d ", temp->key);
        temp = temp->right;
    }
    while (temp != min);
    printf("\n");
}

int main()
{
    insert(10);
    insert(3);
    insert(15);
    insert(6);

    display();

    return 0;
}
